File: Algorithms/0589_N-ary_Tree_Preorder_Traversal/Java/N-AryTreePreorderTraversalSolution2.java

// Space: O(1)
// Time: O(n)

/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {

    public List<Integer> preorder(Node root) {
        if (root == null) return new ArrayList<>();
        if (root.children == null) return new ArrayList<Integer>(Arrays.asList(root.val));
        return dfs(root);
    }

    private List<Integer> dfs(Node root) {
        if (root == null) return new ArrayList<>();
        if (root.children == null) return new ArrayList<Integer>(Arrays.asList(root.val));
        List<Integer> res = new ArrayList<>();
        res.add(root.val);
        for (Node c : root.children) {
            res.addAll(dfs(c));
        }
        return res;
    }
}

File: Algorithms/0589_N-ary_Tree_Preorder_Traversal/Python/N-ary_Tree_Preorder_Traversal_Solution_2.py

# Space: O(1)
# Time: O(n)

"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def preorder(self, root):
        if root is None: return []
        if root.children is None: return root.val

        def dfs(root):
            if root is None: return []
            if root.children is None: return [root.val]
            res = [root.val]
            for c in root.children:
                res += dfs(c)
            return res

        return dfs(root)










